### PR TITLE
[ao][sparsity] Post training data sparsifier callback for lightning

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81813
* #80371
* __->__ #80370

Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestPostTrainingCallback```

### CODE REVIEW

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    Hooks implemented:
+        on_validation_start()
+            1. copies the model and attaches it to the sparsifier
+            2. sparsier step() is called
+            3. squashes the mask()
+    """
+
+    def on_validation_start(self, trainer, pl_module) -> None:
+        copied_model = deepcopy(pl_module.model).eval()  # do we really need to put in cpu?
+        self.sparsified = copied_model
```
Comment (HDCharles) [inline torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:51]: why not assign directly

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            Note: data_list arg should be ignored
+
+    Hooks implemented:
+        on_validation_start()
+            1. copies the model and attaches it to the sparsifier
+            2. sparsier step() is called
+            3. squashes the mask()
+    """
+
+    def on_validation_start(self, trainer, pl_module) -> None:
```
Comment (HDCharles) [inline torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:49]: don't think this is the correct place for this, shouldn't it be on_fit_end for PTS?

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import logging
+from torch.ao.sparsity.experimental.data_sparsifier.base_data_sparsifier import SUPPORTED_TYPES
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+def _create_data_sparsifier(data_sparsifier_args, data_sparsifier_type):
```
Comment (HDCharles) [inline torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py:7]: this function feels a little bit redundant if it just takes 2 arguments and calls one with the other as an arg.

if not, should probably have arg order swapped to be more intuitive

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    def run_all_checks(self, sparsifier_args):
+        pl_module = self._get_pl_module()
+        callback = self._get_callback(sparsifier_args)
+
+        self.check_on_validation_start(pl_module, callback, sparsifier_args)
+
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py:81]: modified the test structure

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py (hunk 1 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py]: done

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py (hunk 2 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py]: changed this from to on_fit_end()

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py (hunk 3 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py]: yeah, seems redundant - directly creating the object now

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+from copy import deepcopy
+import torch
+from dataclasses import dataclass, field
+from typing import Any, Optional, Dict
+import pytorch_lightning as pl
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:5]:
```
import pytorch_lightning as pl  # type: ignore
```

**Review by z-a-f — APPROVED** [review]

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    """Lightning callback that enables post-training sparsity.
+
+    This callback aims to sparsify the model inside lightning module after training.
+    **Note that the model is copied and then sparsified, so the existing model is not modified**
+
+    The sparsified model can be used for comparison and can be accessed using
+        <callback_obj>.sparsified
+
+    Args:
+        data_sparsifier_type (some implemented class of BaseDataSparsifier)
```
Comment (jerryzh168) [inline torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:32]: maybe just `data_sparsifier_class`? I feel type is a bit ambiguous

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            The data sparsifier object of this type is created when the
+            training starts.
+            Note: Objects should not be passed in here as they are created
+            once the training completes.
+
+        data_sparsifier_args (Dict)
```
Comment (jerryzh168) [inline torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:38]: wondering if you would like to adopt the `with_args` design: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/observer.py#L1460

Comment (jerryzh168) [reply torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:38]: in this case `data_sparsifier_type` should be named as `data_sparsifier_constructor` I think

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    Hooks implemented:
+        on_fit_end()
+            1. copies the model and attaches it to the sparsifier
+            2. sparsier step() is called
+            3. squashes the mask()
+    """
+
+    def on_fit_end(self, trainer, pl_module) -> None:
+        self.sparsified = deepcopy(pl_module.model).eval()
+        self.data_sparsifier = self.data_sparsifier_type(**self.data_sparsifier_args)
```
Comment (jerryzh168) [inline torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:51]: also I'm wondering if we can just pass in a data_sparsifier directly from user?

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py (hunk 5 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py]: This still seems to throw TYPEIGNORE error. I get this 
```This line has an unqualified `type: ignore`; please convert it to `type:
    ignore[xxxx]```

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py (hunk 6 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py]: Yeah, makes sense. Changed it to data_sparsifier_class for now.

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py (hunk 8 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py]: We could do that for the data_sparsifier, but I don't think we can for the scheduler. Because the data_scheduler is created after the sparsifier is attached to the model.
So, just to keep things simple, we ask the user to input only the class_name and not the object for both data_sparsifier and data_scheduler.

File: torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py (hunk 5 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py]: ```type: ignore[import]``` seemed to fix the issue.

File: torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            i = c
+
+
+class DummyLightningModule(pl.LightningModule):
+    def __init__(self, iC: int, oC: List[int]):
+        super().__init__()
+        self.model = DummyModel(iC, oC)
+
+    def forward(self):
+        pass
```
Comment (z-a-f) [inline torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py:33]: Let's wrap this into a creator function to encapsulate the pl import? Something like:

```
def make_lightning_module(*args, **kwargs):
    import pytorch_lightning as pl # type: ignore[import]
    class DummyLightningModule(pl.LightningModule):
        ...
    return DummyLightningModule(*args, **kwargs)
```

File: torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+class TestPostTrainingCallback(_PostTrainingCallbackTestCase):
+    def test_post_training_callback(self):
+        sparsifier_args = {
+            'sparsity_level': 0.5,
+            'sparse_block_shape': (1, 4),
+            'zeros_per_block': 4
+        }
+        callback = PostTrainingDataSparsity(DataNormSparsifier, sparsifier_args)
+        pl_module = DummyLightningModule(100, [128, 256, 16])
```
Comment (z-a-f) [inline torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py:83]: If using a creator function, you can replace this with

```
pl_module = make_lightning_module(100, [128, 256, 16])
```

File: torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+            # make sure that non-zero values in data after squash mask are equal to original values
+            sparsified_data = callback.data_sparsifier.get_data(name=valid_name, return_original=False)
+            assert torch.all(sparsified_data[sparsified_data != 0] == param[sparsified_data != 0])
+
+    def run_all_checks(self, pl_module, callback, sparsifier_args):
+        self._check_on_fit_end(pl_module, callback, sparsifier_args)
+
+
+class TestPostTrainingCallback(_PostTrainingCallbackTestCase):
+    def test_post_training_callback(self):
```
Comment (z-a-f) [inline torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py:76]: Can we add a decorator to skip? Something like

```
@unittest.skipIf(not importlib.util.find_spec("pytorch_lightning"), "No pytorch_lightning")
def test_post_training_callback(self):
    ...
```

Double check if the skipping works :)) I didn't confirm lolol

**Review by z-a-f — CHANGES_REQUESTED** [review]

File: torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py (hunk 9 above)
Comment (macandro96) [reply torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py]: yep

File: torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py (hunk 11 above)
Comment (macandro96) [reply torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py:81]: I did - seems like all the tests pass now

File: torch/ao/sparsity/_experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def _attach_model_to_data_sparsifier(module, data_sparsifier):
+    """Attaches a data sparsifier to all the layers of the module.
+    Essentialy, loop over all the weight parameters in the module and
+    attach it to the data sparsifier.
+    Note::
+        The '.' in the layer names are replaced with some other character
+        before attaching to the sparsifier. This is because, the data
+        sparsifier uses a dummy model inside to store the weight parameters.
```
Comment (z-a-f) [inline torch/ao/sparsity/_experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py:14]: Be more specific: what do you mean "with some other character"?

Comment (z-a-f) [reply torch/ao/sparsity/_experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py:14]: Also, I don't think a layer name can have a `.` in there. Do you support FQNs?

File: torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+            super().__init__()
+            self.model = DummyModel(iC, oC)
+
+        def forward(self):
+            pass
+
+    return DummyLightningModule(iC, oC)
+
+
+class _PostTrainingCallbackTestCase(TestCase):
```
Comment (z-a-f) [inline torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py:40]: I still don't see the point of creating a separate class for the test case in this scenario

File: torch/ao/sparsity/_experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py (hunk 12 above)
Comment (macandro96) [reply torch/ao/sparsity/_experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py:14]: Yes, the layer name cannot have a ```.```. We're basically taking the name of the model parameter and replacing it with ```_```.

**Review by z-a-f — APPROVED** [review]
